### PR TITLE
WebSocket options prep

### PR DESCRIPTION
**Pull request by Shane32**
Requires #799

### CODE REVIEW

**Comment by sungam3r** [comment]
Will review after #799 being merged.

File: src/Transports.AspNetCore/Extensions/GraphQLBuilderUserContextExtensions.cs
Comment (Shane32) [inline src/Transports.AspNetCore/Extensions/GraphQLBuilderUserContextExtensions.cs:52]:
```suggestion
```

**Comment by Shane32** [comment]
@sungam3r ETA on review?

**Comment by sungam3r** [comment]
It seems that there are not many changes. I think I will review tomorrow morning.

**Comment by Shane32** [comment]
Sounds good; I hope you are having a good vacation.  I should be able to put the rest of my changes to WebSocket support into a PR after this is merged (I think).  Not sure if I'll be able to add tests into a PR without everything else being merged.  It will probably make more sense just to wait until all the PRs are merged and then add the tests in.

**Comment by Shane32** [comment]
Oh, and then documentation.

**Comment by sungam3r** [comment]
> I hope you are having a good vacation

I wake up at 6AM, go to bed after midnight working outdoors all day long. GitHub activity 2-3 hours at the morning and 1-2 at night.

File: src/Transports.AspNetCore/IWebSocketHandler.cs
Comment (sungam3r) [inline src/Transports.AspNetCore/IWebSocketHandler.cs:19]: Do I understand correctly, that once being accepted WebSocket object is kept alive for ages handling all messages flowing between client and server throught it?

File: src/Transports.AspNetCore/IWebSocketHandler.cs
Comment (sungam3r) [inline src/Transports.AspNetCore/IWebSocketHandler.cs:24]:
```suggestion
    /// Gets a list of supported by this handler WebSocket sub-protocols.
```

File: src/Transports.AspNetCore/WebSockets/GraphQLWebSocketOptions.cs
Comment (sungam3r) [inline src/Transports.AspNetCore/WebSockets/GraphQLWebSocketOptions.cs:10]: Where?

File: src/Transports.AspNetCore/WebSockets/GraphQLWebSocketOptions.cs
Comment (sungam3r) [inline src/Transports.AspNetCore/WebSockets/GraphQLWebSocketOptions.cs:16]:
```suggestion
    /// The default is <see langword="null"/> that means keep-alive packets are disabled.
```

File: src/Transports.AspNetCore/WebSockets/GraphQLWebSocketOptions.cs
Comment (sungam3r) [inline src/Transports.AspNetCore/WebSockets/GraphQLWebSocketOptions.cs:38]:
```suggestion
    /// Disconnects a subscription from the client in case of any GraphQL errors during a subscription.
```

File: src/Transports.AspNetCore/GraphQLHttpMiddleware.cs
Comment (sungam3r) [inline src/Transports.AspNetCore/GraphQLHttpMiddleware.cs:43]: The first thought looking at this signature - why not typed handler - `IWebSocketHandler<TSchema>`?

File: src/Transports.AspNetCore/GraphQLHttpMiddleware.cs
Comment (sungam3r) [inline src/Transports.AspNetCore/GraphQLHttpMiddleware.cs:200]: 🤔 It looks a bit weird to wrap existing method into `UserContextBuilder` object when inside `BuildUserContextAsync` you pull `IUserContextBuilder` from DI. Why not use `context.RequestServices.GetService<IUserContextBuilder>()` for `await selectedHandler.ExecuteAsync` line?

File: src/Transports.AspNetCore/GraphQLHttpMiddleware.cs
Comment (sungam3r) [inline src/Transports.AspNetCore/GraphQLHttpMiddleware.cs:207]: I'd remove `// Process WebSocket request `comment  below since `HandleWebSocketAsync` is pretty obvios and also has xml comments.

File: src/Transports.AspNetCore/GraphQLHttpMiddleware.cs
Comment (sungam3r) [inline src/Transports.AspNetCore/GraphQLHttpMiddleware.cs:523]: Search all `await next` lines across the repo. I see 3 usages without `ConfigureAwait(false)`.

File: src/Transports.AspNetCore/GraphQLHttpMiddleware.cs
Comment (sungam3r) [inline src/Transports.AspNetCore/GraphQLHttpMiddleware.cs:571]: Regarding `  if (socket.SubProtocol != selectedProtocol)` conditions. Is such a situation possible in reality?

**Review by sungam3r — COMMENTED** [review]
I looked through all files. So far, I understand the basic behavior, although there is no actual implementation yet and I do not understand anything in WebSockets.

File: src/Transports.AspNetCore/IWebSocketHandler.cs
Comment (Shane32) [reply src/Transports.AspNetCore/IWebSocketHandler.cs:19]: Yes that is how ASP.NET Core handles WebSocket connections. When the middleware returns, the associated WebSocket connection/object is automatically closed. So the middleware won’t return until the WebSocket connection has disconnected, or until it wants to disconnect.

File: src/Transports.AspNetCore/IWebSocketHandler.cs
Comment (Shane32) [reply src/Transports.AspNetCore/IWebSocketHandler.cs:24]:
```suggestion
    /// Gets a list of WebSocket sub-protocols supported by this handler.
```

File: src/Transports.AspNetCore/WebSockets/GraphQLWebSocketOptions.cs
Comment (Shane32) [reply src/Transports.AspNetCore/WebSockets/GraphQLWebSocketOptions.cs:16]:
```suggestion
    /// The default is <see langword="null"/> which means that keep-alive packets are disabled.
```

File: src/Transports.AspNetCore/WebSockets/GraphQLWebSocketOptions.cs
Comment (Shane32) [reply src/Transports.AspNetCore/WebSockets/GraphQLWebSocketOptions.cs:33]:
```suggestion
    /// <see cref="IObserver{T}.OnError(Exception)"/> event.  The default value is <see langword="true"/>.
```

File: src/Transports.AspNetCore/WebSockets/GraphQLWebSocketOptions.cs
Comment (Shane32) [reply src/Transports.AspNetCore/WebSockets/GraphQLWebSocketOptions.cs]: > How it works? Is it assumed that a subscription can sometimes work even if there are errors? By the way, what errors?

It is allowed for a GraphQL subscription to periodically return data in the form of a properly formatted ExecutionResult (for the subscribed field only). Also it can return errors in the same manner. 

So let’s say that a user subscribed to a notifications about new high scores for a game. Let’s days they requested Score and Name. If both are non null fields, but the data source generating the source events gave a null value for Name, the DocumentExecuter would generate a proper ExecutionResult with an error message saying that a null value was supplied for a non null field. This result can be passed back through the WebSocket connection to the client without breaking the WebSocket connection or subscription. However, it may be desired that for some unanticipated errors, the connection is purposely dropped, forcing the client to reconnect, which may reset whatever the fault was on the server end.

Comment (Shane32) [reply src/Transports.AspNetCore/WebSockets/GraphQLWebSocketOptions.cs]: Note that all validation checks occur before the subscription is completely set up.

File: tests/ApiApprovalTests/GraphQL.Server.Transports.AspNetCore.approved.txt
Comment (Shane32) [reply tests/ApiApprovalTests/GraphQL.Server.Transports.AspNetCore.approved.txt:155]: In this case the name of the WebSocket message (per the protocol) is ConnectionInit - hence the name. But we can discuss if another naming strategy is better.

File: src/Transports.AspNetCore/GraphQLHttpMiddleware.cs
Comment (Shane32) [reply src/Transports.AspNetCore/GraphQLHttpMiddleware.cs:43]: The “base” class uses untyped IDocumentExecuter and untyped IWebSocketHandler. This ctor isn’t finished because the supporting class isn’t done yet.

File: src/Transports.AspNetCore/GraphQLHttpMiddleware.cs
Comment (Shane32) [reply src/Transports.AspNetCore/GraphQLHttpMiddleware.cs:200]: Because I want it to call the protected method within this class.

File: src/Transports.AspNetCore/GraphQLHttpMiddleware.cs
Comment (Shane32) [reply src/Transports.AspNetCore/GraphQLHttpMiddleware.cs:523]: There is not a single use of ConfigureAwait(false) in this project. Probably because it generally isn’t necessary on ASP.NET Core. I thought we would add that to all code but I want to finish all my merges first.

Comment (Shane32) [reply src/Transports.AspNetCore/GraphQLHttpMiddleware.cs:523]: There will probably be 100s of references that need to be fixed.

File: src/Transports.AspNetCore/GraphQLHttpMiddleware.cs
Comment (Shane32) [reply src/Transports.AspNetCore/GraphQLHttpMiddleware.cs:571]: I don’t know actually. It seems likely not but this validation check was in the old code and I didn’t see a good reason to remove it.

**Comment by Shane32** [comment]
> I looked through all files. So far, I understand the basic behavior, although there is no actual implementation yet and I do not understand anything in WebSockets.

I hope and think that you will find it easy to follow the logic in my PRs and will understand how subscriptions work to some degree by the time you’ve reviewed all my PRs.

File: src/Transports.AspNetCore/WebSockets/GraphQLWebSocketOptions.cs
Comment (Shane32) [reply src/Transports.AspNetCore/WebSockets/GraphQLWebSocketOptions.cs:27]:
```suggestion
    /// The default is typically 10 seconds.
```
Changing to this comment.

File: tests/ApiApprovalTests/GraphQL.Server.Transports.AspNetCore.approved.txt
Comment (sungam3r) [reply tests/ApiApprovalTests/GraphQL.Server.Transports.AspNetCore.approved.txt:155]: Understood. Keep as is.

File: src/Transports.AspNetCore/GraphQLHttpMiddleware.cs
Comment (sungam3r) [reply src/Transports.AspNetCore/GraphQLHttpMiddleware.cs:523]: > There is not a single use of ConfigureAwait(false) in this project.

изображение

Comment (Shane32) [reply src/Transports.AspNetCore/GraphQLHttpMiddleware.cs:523]: Not many then.

Comment (Shane32) [reply src/Transports.AspNetCore/GraphQLHttpMiddleware.cs:523]: lol - that line was one you added

Comment (Shane32) [reply src/Transports.AspNetCore/GraphQLHttpMiddleware.cs:523]: For consistency, I should remove it...

**Review by sungam3r — APPROVED** [review]

**Pull request merged**
